pkg/store/store.go

```diff
@@ -49,11 +51,19 @@ type Store interface {
 
 // Key is a key for the object store.
 type Key struct {
-	Namespace  string      `json:"namespace"`
-	APIVersion string      `json:"apiVersion"`
-	Kind       string      `json:"kind"`
-	Name       string      `json:"name"`
-	Selector   *labels.Set `json:"selector"`
+	Namespace            string         `json:"namespace"`
+	APIVersion           string         `json:"apiVersion"`
+	Kind                 string         `json:"kind"`
+	Name                 string         `json:"name"`
+	Selector             *labels.Set    `json:"selector"`
+	RequirementsSelector *[]Requirement `json:"requirementsSelector"`
+}
+
+// See vendor/k8s.io/apimachinery/pkg/labels/selector.go:120
```

Comment:
Why duplicate this here? We already import labels. Is it just so we can annotate it?

Reply:
All the fields in `labels#Requirement` are (a) private, and (b) not annotated, and it _seemed_ to have issues when I tried it. I'm happy to give it another go, though.

Reply:
Yeah, just checked again, and it silently ignores the values in the JSON - it creates an array with a single object, but all the values in that object are `nil`/`""`/etc. Unsure if that's because they're private or because they're not annotated?

Reply:
> All the fields in labels#Requirement are (a) private, and (b) not annotated, and it seemed to have issues when I tried it.

That's usually a sign that you're trying to use sometime in a way it wasn't intended to be used.

Have you looked at https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#LabelSelector

Reply:
That looks like it would work, but it's not a particularly nice API, as there are two different ways of searching for `key=value`, one of which duplicates what we already have in `Key#Selector`.

`equals`:
```
{
  "matchLabels": { "label1":"test" }
}
```
OR
```
{
  "matchExpressions": [{"key":"label1", "operator":"in", "values":["test"]}]
}
```

`exists`:
```
{
  "matchExpressions": [{"key":"label1", "operator":"exists", "values":[]}]
}
```

Reply:
It also doesn't support the full set of operators, missing greater-than and less-than operators, though I'm not sure how much use they'd see in practice?

Reply:
I've implemented it with that struct, and I can push that out if you're interested in seeing it. Quirks:

* Have to write our own `String()` function for it
* Have to convert *both* `matchLabels` and `matchExpressions` to a String, which produces an awkward result like `CacheKey[Namespace='default', APIVersion='v1', Kind='ConfigMap', LabelSelector='MatchLabels=foo1=bar1,foo4=bar4, MatchExpressions=foo2 In (bar2),foo3 Exists']`

Conversion is at least handled by `metav1.LabelSelectorAsSelector`, so apart from the API design and the `String()` function, it's pretty simple.

Reply:
The core types are not always the best example for good greenfield design , but they are well understood in the community. https://godoc.org/k8s.io/api/apps/v1#DeploymentSpec

> It also doesn't support the full set of operators, missing greater-than and less-than operators

Greater than and less than operators on strings are always a bit dodgy unless you also define the character set, locale,  lexicographical rules, etc.

Reply:
That's a fair point. I've pushed a commit using `LabelSelector` instead of the custom type. Might still be some tweaks needed, especially to the `String` function.